File: backend/apps/core/money.py
"""
Money arithmetic — the single authority on how an order total is computed.

This module is deliberately free of Django imports. It is pure Decimal logic so
the exact same algorithm can be vendored into the PySide6 Desktop client, and
both sides can be validated against the same golden-file fixture
(tests/fixtures/money_cases.json). If the Desktop and the server ever disagree
by one piaster, CI fails.

Rules (docs/02-data-model.md):
  * Decimal only. float never touches money.
  * ROUND_HALF_UP, applied once at each documented boundary — never mid-calculation.
  * The order of operations below is fixed and must not be reordered.

      line_gross     = round(unit_price_with_modifiers × quantity)
      line_discount  = round(line_gross × line_discount_pct / 100)
      line_net       = line_gross − line_discount
      order_net      = Σ line_net
      order_discount = round(order_net × order_discount_pct / 100)
      after_discount = order_net − order_discount
      service        = round(after_discount × service_pct / 100)
      tax            = f(after_discount, service, taxable_ratio, vat_pct, inclusive)
      grand_total    = after_discount + service + tax   (+ rounding adjustment)
"""

from __future__ import annotations

from dataclasses import dataclass, field
from decimal import ROUND_HALF_UP, Decimal

__all__ = [
    "ZERO",
    "LineTotals",
    "OrderLine",
    "OrderTotals",
    "TaxRules",
    "compute_line",
    "compute_order",
    "quantize_money",
    "round_to_step",
]

ZERO = Decimal("0.00")
CENT = Decimal("0.01")
HUNDRED = Decimal("100")


def quantize_money(value: Decimal) -> Decimal:
    """Round to 2 decimal places, half away from zero."""
    return value.quantize(CENT, rounding=ROUND_HALF_UP)


def round_to_step(value: Decimal, step: Decimal) -> Decimal:
    """
    Round a money value to the nearest multiple of `step`.

    Used for `finance.rounding_step` — a cafe that has abolished small coins can
    set 0.25 and have every total land on a quarter pound. A step of 0.01 (the
    default) is a no-op beyond normal money quantization.
    """
    if step <= ZERO:
        return quantize_money(value)
    steps = (value / step).quantize(Decimal("1"), rounding=ROUND_HALF_UP)
    return quantize_money(steps * step)


@dataclass(frozen=True)
class TaxRules:
    """
    Resolved tax configuration for one order.

    Every field comes from the settings registry (docs/11-configuration.md).

    There are deliberately NO defaults here. A default would be a second source
    of truth for the VAT rate, and a service that constructed `TaxRules()` by
    mistake would silently charge 14% instead of the branch's configured rate.
    Requiring every field means that mistake cannot compile.
    """

    vat_percent: Decimal
    vat_enabled: bool
    vat_inclusive: bool
    service_percent: Decimal
    service_enabled: bool
    rounding_step: Decimal

    @property
    def effective_vat(self) -> Decimal:
        return self.vat_percent if self.vat_enabled else ZERO

    @property
    def effective_service(self) -> Decimal:
        return self.service_percent if self.service_enabled else ZERO


@dataclass(frozen=True)
class OrderLine:
    """One sellable line before totals are computed."""

    unit_price: Decimal
    quantity: Decimal = Decimal("1")
    discount_percent: Decimal = ZERO
    modifier_deltas: tuple[Decimal, ...] = field(default_factory=tuple)
    tax_exempt: bool = False

    @property
    def effective_unit_price(self) -> Decimal:
        return self.unit_price + sum(self.modifier_deltas, ZERO)


@dataclass(frozen=True)
class LineTotals:
    gross: Decimal
    discount: Decimal
    net: Decimal


@dataclass(frozen=True)
class OrderTotals:
    subtotal: Decimal
    """Sum of line nets, before any order-level discount."""
    discount_total: Decimal
    """Line discounts plus the order-level discount."""
    service_total: Decimal
    tax_total: Decimal
    grand_total: Decimal
    rounding_adjustment: Decimal
    """Difference introduced by `rounding_step`. Reported, never hidden."""
    lines: tuple[LineTotals, ...] = ()


def compute_line(line: OrderLine) -> LineTotals:
    gross = quantize_money(line.effective_unit_price * line.quantity)
    discount = quantize_money(gross * line.discount_percent / HUNDRED)
    return LineTotals(gross=gross, discount=discount, net=gross - discount)


def compute_order(
    lines: list[OrderLine] | tuple[OrderLine, ...],
    rules: TaxRules,
    order_discount_percent: Decimal = ZERO,
) -> OrderTotals:
    """
    Compute authoritative totals for an order.

    Tax on a mixed order is apportioned: when some lines are tax-exempt, VAT
    applies to the taxable share of the post-discount amount (service charge
    included, since service is itself taxable). Apportioning by value is the
    defensible reading and keeps a fully-exempt or fully-taxable order exact.
    """
    line_totals = tuple(compute_line(line) for line in lines)

    order_net = sum((lt.net for lt in line_totals), ZERO)
    line_discount_total = sum((lt.discount for lt in line_totals), ZERO)

    order_discount = quantize_money(order_net * order_discount_percent / HUNDRED)
    after_discount = order_net - order_discount

    service = quantize_money(after_discount * rules.effective_service / HUNDRED)

    taxable_net = sum(
        (lt.net for lt, line in zip(line_totals, lines, strict=True) if not line.tax_exempt),
        ZERO,
    )
    taxable_ratio = (taxable_net / order_net) if order_net else Decimal("0")

    tax = _compute_tax(
        base=after_discount + service,
        taxable_ratio=taxable_ratio,
        vat_percent=rules.effective_vat,
        inclusive=rules.vat_inclusive,
    )

    if rules.vat_inclusive:
        # Prices already contain VAT, so tax is disclosed rather than added.
        raw_total = after_discount + service
    else:
        raw_total = after_discount + service + tax

    grand_total = round_to_step(raw_total, rules.rounding_step)

    return OrderTotals(
        subtotal=quantize_money(order_net),
        discount_total=quantize_money(line_discount_total + order_discount),
        service_total=service,
        tax_total=tax,
        grand_total=grand_total,
        rounding_adjustment=quantize_money(grand_total - raw_total),
        lines=line_totals,
    )


def _compute_tax(
    *,
    base: Decimal,
    taxable_ratio: Decimal,
    vat_percent: Decimal,
    inclusive: bool,
) -> Decimal:
    if vat_percent <= ZERO or base <= ZERO or taxable_ratio <= ZERO:
        return ZERO

    taxable_base = quantize_money(base * taxable_ratio)

    if inclusive:
        # Extract the VAT already contained in the price.
        net_of_tax = taxable_base / (Decimal("1") + vat_percent / HUNDRED)
        return quantize_money(taxable_base - net_of_tax)

    return quantize_money(taxable_base * vat_percent / HUNDRED)

File: frontend/src/lib/money.ts
/**
 * Money arithmetic for the browser — the till's authority on an order total.
 *
 * **Why this file exists at all.** The Desktop client is cancelled, so the
 * browser is now the offline point of sale. An offline till has to compute the
 * total it prints on a receipt and hands to a customer, which means the
 * arithmetic has to exist here — it cannot be a server round trip when the whole
 * point is that the server is unreachable.
 *
 * **Why it is a PORT and not a vendored copy.** `apps/core/money.py` is copied
 * verbatim into the Desktop by `scripts/vendor_shared.py`, and `--check` fails CI
 * if the copies drift. That is the right discipline and it is not available here:
 * there is no artefact to vendor between Python and a browser. The project has
 * already met this exact problem once — the floor geometry exists twice, in
 * TypeScript for CSS and in Python for QPainter, "because there is no artefact to
 * vendor between a browser and Qt", and is tested twice against the same cases.
 *
 * So this follows that precedent, and the safeguard is the same one: **the golden
 * fixture is read across the monorepo, not copied.** `money.test.ts` runs
 * `backend/tests/fixtures/money_cases.json` — the server's own file, the same one
 * `test_money_golden.py` runs — through this implementation. A copy of a fixture
 * is a thing that drifts; a shared read is not.
 *
 * **float never touches money.** `decimal.js`, not `number`. This is the reason
 * the dependency is here rather than integer piasters: the inclusive-VAT path
 * divides by `1 + rate`, and expressing that in integers either loses precision
 * or reimplements a decimal library badly.
 *
 * The order of operations below is fixed and mirrors money.py line for line. Do
 * not reorder it; the fixture was hand-computed against this sequence.
 *
 *     line_gross     = round(unit_price_with_modifiers × quantity)
 *     line_discount  = round(line_gross × line_discount_pct / 100)
 *     line_net       = line_gross − line_discount
 *     order_net      = Σ line_net
 *     order_discount = round(order_net × order_discount_pct / 100)
 *     after_discount = order_net − order_discount
 *     service        = round(after_discount × service_pct / 100)
 *     tax            = f(after_discount, service, taxable_ratio, vat_pct, inclusive)
 *     grand_total    = after_discount + service + tax   (+ rounding adjustment)
 */
import Decimal from 'decimal.js'

/**
 * 28 significant digits, matching Python's default Decimal context.
 *
 * Not cosmetic. `net_of_tax = taxable_base / (1 + vat/100)` is a division whose
 * intermediate precision decides the piaster: decimal.js defaults to 20 digits
 * and Python to 28, and a fixture case can land either side of a half-piaster
 * boundary between the two. Set once, globally, because a per-call precision is
 * one call site away from being forgotten.
 *
 * `ROUND_HALF_UP` is the library default for `toDP`, but stated anyway — this is
 * the rounding mode docs/02 specifies and the one the golden file was computed
 * with, and a default that goes unstated is a default somebody changes.
 */
Decimal.set({ precision: 28, rounding: Decimal.ROUND_HALF_UP })

export const ZERO = new Decimal('0.00')
const HUNDRED = new Decimal(100)

/** A money value, as a string. Strings in, strings out — see `money.py`. */
export type Money = string

export interface TaxRules {
  vat_percent: Money
  vat_enabled: boolean
  vat_inclusive: boolean
  service_percent: Money
  service_enabled: boolean
  rounding_step: Money
}

export interface OrderLine {
  unit_price: Money
  quantity?: Money
  discount_percent?: Money
  modifier_deltas?: Money[]
  tax_exempt?: boolean
}

export interface LineTotals {
  gross: Money
  discount: Money
  net: Money
}

export interface OrderTotals {
  subtotal: Money
  discount_total: Money
  service_total: Money
  tax_total: Money
  grand_total: Money
  /** The difference `rounding_step` introduced. Reported, never hidden. */
  rounding_adjustment: Money
  lines: LineTotals[]
}

/** Round to 2 decimal places, half away from zero. */
export function quantizeMoney(value: Decimal): Decimal {
  return value.toDecimalPlaces(2, Decimal.ROUND_HALF_UP)
}

/**
 * Round to the nearest multiple of `step`.
 *
 * `finance.rounding_step` — a cafe that has abolished small coins sets 0.25 and
 * every total lands on a quarter pound. A step of 0.01 is a no-op beyond normal
 * quantization.
 */
export function roundToStep(value: Decimal, step: Decimal): Decimal {
  if (step.lte(0)) return quantizeMoney(value)
  const steps = value.div(step).toDecimalPlaces(0, Decimal.ROUND_HALF_UP)
  return quantizeMoney(steps.mul(step))
}

function effectiveVat(rules: TaxRules): Decimal {
  return rules.vat_enabled ? new Decimal(rules.vat_percent) : ZERO
}

function effectiveService(rules: TaxRules): Decimal {
  return rules.service_enabled ? new Decimal(rules.service_percent) : ZERO
}

function effectiveUnitPrice(line: OrderLine): Decimal {
  return (line.modifier_deltas ?? []).reduce(
    (total, delta) => total.plus(new Decimal(delta)),
    new Decimal(line.unit_price),
  )
}

export function computeLine(line: OrderLine): LineTotals {
  const gross = quantizeMoney(effectiveUnitPrice(line).mul(new Decimal(line.quantity ?? '1')))
  const discount = quantizeMoney(gross.mul(new Decimal(line.discount_percent ?? '0')).div(HUNDRED))
  return {
    gross: gross.toFixed(2),
    discount: discount.toFixed(2),
    net: gross.minus(discount).toFixed(2),
  }
}

/**
 * Tax on a mixed order is APPORTIONED.
 *
 * When some lines are exempt, VAT applies to the taxable share of the
 * post-discount amount — service included, because service is itself taxable.
 * Apportioning by value is the defensible reading and keeps a fully-exempt or
 * fully-taxable order exact.
 */
function computeTax(
  base: Decimal,
  taxableRatio: Decimal,
  vatPercent: Decimal,
  inclusive: boolean,
): Decimal {
  if (vatPercent.lte(0) || base.lte(0) || taxableRatio.lte(0)) return ZERO

  const taxableBase = quantizeMoney(base.mul(taxableRatio))

  if (inclusive) {
    // Extract the VAT already contained in the price.
    const netOfTax = taxableBase.div(new Decimal(1).plus(vatPercent.div(HUNDRED)))
    return quantizeMoney(taxableBase.minus(netOfTax))
  }

  return quantizeMoney(taxableBase.mul(vatPercent).div(HUNDRED))
}

export function computeOrder(
  lines: OrderLine[],
  rules: TaxRules,
  orderDiscountPercent: Money = '0',
): OrderTotals {
  const lineTotals = lines.map(computeLine)

  const orderNet = lineTotals.reduce((total, lt) => total.plus(new Decimal(lt.net)), ZERO)
  const lineDiscountTotal = lineTotals.reduce(
    (total, lt) => total.plus(new Decimal(lt.discount)),
    ZERO,
  )

  const orderDiscount = quantizeMoney(orderNet.mul(new Decimal(orderDiscountPercent)).div(HUNDRED))
  const afterDiscount = orderNet.minus(orderDiscount)

  const service = quantizeMoney(afterDiscount.mul(effectiveService(rules)).div(HUNDRED))

  const taxableNet = lineTotals.reduce(
    (total, lt, index) => (lines[index].tax_exempt ? total : total.plus(new Decimal(lt.net))),
    ZERO,
  )
  const taxableRatio = orderNet.isZero() ? ZERO : taxableNet.div(orderNet)

  const tax = computeTax(
    afterDiscount.plus(service),
    taxableRatio,
    effectiveVat(rules),
    rules.vat_inclusive,
  )

  // Inclusive prices already contain VAT, so tax is disclosed rather than added.
  const rawTotal = rules.vat_inclusive
    ? afterDiscount.plus(service)
    : afterDiscount.plus(service).plus(tax)

  const grandTotal = roundToStep(rawTotal, new Decimal(rules.rounding_step))

  return {
    subtotal: quantizeMoney(orderNet).toFixed(2),
    discount_total: quantizeMoney(lineDiscountTotal.plus(orderDiscount)).toFixed(2),
    service_total: service.toFixed(2),
    tax_total: tax.toFixed(2),
    grand_total: grandTotal.toFixed(2),
    rounding_adjustment: quantizeMoney(grandTotal.minus(rawTotal)).toFixed(2),
    lines: lineTotals,
  }
}
